File: apps/gateway/internal/plugins/manifest.go
// Package plugins provides manifest parsing, validation, and permission
// enforcement for the UBAG WASM plugin host.
//
// The Go types here mirror the TypeScript definitions in
// packages/plugins/src/manifest.ts.  ParseManifest is the single entry point:
// it deserialises raw JSON and returns a fully-validated Manifest or a
// ValidationError whose Issues slice contains every problem found.
package plugins

import (
	"encoding/json"
	"fmt"
	"regexp"
	"strings"
)

// SchemaVersion is the only accepted value for the schema_version field.
const SchemaVersion = "ubag.plugin.v0"

// Capability names.
type Capability string

const (
	CapabilityTransformPrompt   Capability = "transform.prompt"
	CapabilityTransformResponse Capability = "transform.response"
	CapabilityHookJobPre        Capability = "hook.job.pre"
	CapabilityHookJobPost       Capability = "hook.job.post"
)

var knownCapabilities = map[Capability]struct{}{
	CapabilityTransformPrompt:   {},
	CapabilityTransformResponse: {},
	CapabilityHookJobPre:        {},
	CapabilityHookJobPost:       {},
}

// HostFunctionName represents a named host function a plugin may request.
type HostFunctionName string

const (
	HostFnLog     HostFunctionName = "log"
	HostFnClock   HostFunctionName = "clock"
	HostFnRandom  HostFunctionName = "random"
	HostFnFetch   HostFunctionName = "fetch"
	HostFnReadFile HostFunctionName = "read_file"
	HostFnGetEnv  HostFunctionName = "get_env"
)

var knownHostFunctions = map[HostFunctionName]struct{}{
	HostFnLog:     {},
	HostFnClock:   {},
	HostFnRandom:  {},
	HostFnFetch:   {},
	HostFnReadFile: {},
	HostFnGetEnv:  {},
}

// EntrypointType identifies the WASM component model variant.
type EntrypointType string

const (
	EntrypointWASIComponent EntrypointType = "wasi-component"
	EntrypointWASICommand   EntrypointType = "wasi-command"
	EntrypointCoreModule    EntrypointType = "core-module"
)

// EngineRuntime selects the WASM execution engine.
type EngineRuntime string

const (
	RuntimeWASIPreview1 EngineRuntime = "wasi-preview1"
	RuntimeWASIPreview2 EngineRuntime = "wasi-preview2"
	RuntimeCore         EngineRuntime = "core"
)

// -----------------------------------------------------------------------
// Struct types
// -----------------------------------------------------------------------

// NetworkPermission controls outbound HTTP/HTTPS access.
type NetworkPermission struct {
	Allowed      bool
	AllowedHosts []string
}

// FilesystemPermission controls read access to the host filesystem.
type FilesystemPermission struct {
	Allowed      bool
	AllowedPaths []string
}

// EnvPermission controls access to environment variables.
type EnvPermission struct {
	Allowed     bool
	AllowedKeys []string
}

// Permissions is the full permission set declared in a manifest.
type Permissions struct {
	HostFunctions  []HostFunctionName
	Network        NetworkPermission
	Filesystem     FilesystemPermission
	Env            EnvPermission
	MaxMemoryBytes int64
	MaxExecutionMS int64
}

// Entrypoint describes how the WASM module is loaded and called.
type Entrypoint struct {
	Type    EntrypointType
	Module  string
	Exports EntrypointExports
}

// EntrypointExports names the exported functions within the WASM module.
type EntrypointExports struct {
	Transform string // optional
	Hook      string // optional
	Init      string // optional
}

// Engine describes runtime requirements.
type Engine struct {
	Runtime        EngineRuntime
	MinHostVersion string // optional; "" means unset
}

// Manifest is a validated, fully-typed plugin manifest.
type Manifest struct {
	SchemaVersion string
	ID            string
	DisplayName   string
	Version       string
	Description   string // optional; "" means unset
	Capabilities  []Capability
	Entrypoint    Entrypoint
	Permissions   Permissions
	Engine        Engine
}

// -----------------------------------------------------------------------
// ValidationError
// -----------------------------------------------------------------------

// ValidationError is returned by ParseManifest when one or more fields are
// invalid.  Issues contains one human-readable string per problem.
type ValidationError struct {
	Issues []string
}

func (e *ValidationError) Error() string {
	return fmt.Sprintf("plugin manifest invalid: %s", strings.Join(e.Issues, "; "))
}

// -----------------------------------------------------------------------
// Compiled patterns
// -----------------------------------------------------------------------

var (
	idPattern     = regexp.MustCompile(`^[a-z0-9][a-z0-9_-]{1,63}$`)
	semverPattern = regexp.MustCompile(`^[0-9]+\.[0-9]+\.[0-9]+([-+].+)?$`)
	strictSemver  = regexp.MustCompile(`^[0-9]+\.[0-9]+\.[0-9]+$`)
)

// -----------------------------------------------------------------------
// Constants for defaults and limits
// -----------------------------------------------------------------------

const (
	defaultMaxMemoryBytes int64 = 16_777_216
	defaultMaxExecutionMS int64 = 1_000
	minMemoryBytes        int64 = 65_536
	minExecutionMS        int64 = 1
)

// -----------------------------------------------------------------------
// ParseManifest
// -----------------------------------------------------------------------

// ParseManifest parses and validates a JSON-encoded plugin manifest.
// It returns a Manifest on success or a *ValidationError that lists every
// problem found.
func ParseManifest(data []byte) (Manifest, error) {
	// Step 1: JSON decode into a generic map.
	var raw any
	if err := json.Unmarshal(data, &raw); err != nil {
		return Manifest{}, &ValidationError{Issues: []string{
			fmt.Sprintf("manifest is not valid JSON: %v", err),
		}}
	}

	record, ok := raw.(map[string]any)
	if !ok {
		return Manifest{}, &ValidationError{Issues: []string{
			"manifest must be a JSON object",
		}}
	}

	issues := &issueCollector{}

	// schema_version
	if sv, _ := record["schema_version"].(string); sv != SchemaVersion {
		issues.add(fmt.Sprintf("schema_version must be %q", SchemaVersion))
	}

	// id
	id, _ := record["id"].(string)
	if !idPattern.MatchString(id) {
		issues.add("id must match ^[a-z0-9][a-z0-9_-]{1,63}$")
	}

	// display_name
	displayName, _ := record["display_name"].(string)
	if displayName == "" {
		issues.add("display_name must be a non-empty string")
	}

	// version
	version, _ := record["version"].(string)
	if !semverPattern.MatchString(version) {
		issues.add("version must be a semantic version string")
	}

	// description (optional)
	var description string
	if rawDesc, exists := record["description"]; exists {
		if s, ok := rawDesc.(string); ok {
			description = s
		} else if rawDesc != nil {
			issues.add("description must be a string")
		}
	}

	// capabilities
	caps := parseCapabilities(record["capabilities"], issues)

	// entrypoint
	ep := parseEntrypoint(record["entrypoint"], issues)

	// permissions
	perms := parsePermissions(record["permissions"], issues)

	// engine
	eng := parseEngine(record["engine"], issues)

	if len(issues.list) > 0 {
		return Manifest{}, &ValidationError{Issues: issues.list}
	}

	return Manifest{
		SchemaVersion: SchemaVersion,
		ID:            id,
		DisplayName:   displayName,
		Version:       version,
		Description:   description,
		Capabilities:  caps,
		Entrypoint:    ep,
		Permissions:   perms,
		Engine:        eng,
	}, nil
}

// -----------------------------------------------------------------------
// Sub-parsers
// -----------------------------------------------------------------------

func parseCapabilities(raw any, issues *issueCollector) []Capability {
	arr, ok := raw.([]any)
	if !ok || len(arr) == 0 {
		issues.add("capabilities must be a non-empty array")
		return nil
	}
	seen := map[Capability]struct{}{}
	out := make([]Capability, 0, len(arr))
	for _, entry := range arr {
		s, ok := entry.(string)
		if !ok {
			issues.addUnique(fmt.Sprintf(`capabilities contains unknown capability "%v"`, entry))
			continue
		}
		cap := Capability(s)
		if _, known := knownCapabilities[cap]; !known {
			issues.addUnique(fmt.Sprintf(`capabilities contains unknown capability "%s"`, s))
			continue
		}
		if _, dup := seen[cap]; !dup {
			seen[cap] = struct{}{}
			out = append(out, cap)
		}
	}
	return out
}

func parseHostFunctions(raw any, issues *issueCollector) []HostFunctionName {
	arr, ok := raw.([]any)
	if !ok {
		issues.add("permissions.host_functions must be an array")
		return nil
	}
	seen := map[HostFunctionName]struct{}{}
	out := make([]HostFunctionName, 0, len(arr))
	for _, entry := range arr {
		s, ok := entry.(string)
		if !ok {
			issues.addUnique(fmt.Sprintf(`permissions.host_functions contains unknown host function "%v"`, entry))
			continue
		}
		fn := HostFunctionName(s)
		if _, known := knownHostFunctions[fn]; !known {
			issues.addUnique(fmt.Sprintf(`permissions.host_functions contains unknown host function "%s"`, s))
			continue
		}
		if _, dup := seen[fn]; !dup {
			seen[fn] = struct{}{}
			out = append(out, fn)
		}
	}
	return out
}

// parsePermissionToggle parses a network / filesystem / env sub-object.
// listKey is the name of the string-array field ("allowed_hosts" etc.)
func parsePermissionToggle(raw any, field, listKey string, issues *issueCollector) (allowed bool, list []string) {
	if raw == nil {
		return false, nil
	}
	rec, ok := raw.(map[string]any)
	if !ok {
		issues.addUnique(fmt.Sprintf("permissions.%s must be an object", field))
		return false, nil
	}

	if b, ok := rec["allowed"].(bool); ok {
		allowed = b
	} else {
		issues.addUnique(fmt.Sprintf("permissions.%s.allowed must be a boolean", field))
	}

	if rawList, exists := rec[listKey]; exists {
		arr, ok := rawList.([]any)
		if !ok {
			issues.addUnique(fmt.Sprintf("permissions.%s.%s must be a string array", field, listKey))
		} else {
			for _, item := range arr {
				s, ok := item.(string)
				if !ok {
					issues.addUnique(fmt.Sprintf("permissions.%s.%s must be a string array", field, listKey))
					break
				}
				list = append(list, s)
			}
		}
	}

	if !allowed && len(list) > 0 {
		issues.addUnique(fmt.Sprintf(
			"permissions.%s.%s is set but permissions.%s.allowed is false",
			field, listKey, field))
	}
	return allowed, list
}

func parsePermissions(raw any, issues *issueCollector) Permissions {
	rec, ok := raw.(map[string]any)
	if !ok {
		issues.addUnique("permissions must be an object")
		rec = map[string]any{}
	}

	hostFns := parseHostFunctions(rec["host_functions"], issues)

	netAllowed, netHosts := parsePermissionToggle(rec["network"], "network", "allowed_hosts", issues)
	fsAllowed, fsPaths := parsePermissionToggle(rec["filesystem"], "filesystem", "allowed_paths", issues)
	envAllowed, envKeys := parsePermissionToggle(rec["env"], "env", "allowed_keys", issues)

	// Cross-checks: capability must imply the required host function.
	if netAllowed && !containsHostFn(hostFns, HostFnFetch) {
		issues.addUnique(`permissions.network.allowed requires host function "fetch"`)
	}
	if fsAllowed && !containsHostFn(hostFns, HostFnReadFile) {
		issues.addUnique(`permissions.filesystem.allowed requires host function "read_file"`)
	}
	if envAllowed && !containsHostFn(hostFns, HostFnGetEnv) {
		issues.addUnique(`permissions.env.allowed requires host function "get_env"`)
	}

	maxMem := parsePositiveInt(rec["max_memory_bytes"], "permissions.max_memory_bytes", defaultMaxMemoryBytes, minMemoryBytes, issues)
	maxExec := parsePositiveInt(rec["max_execution_ms"], "permissions.max_execution_ms", defaultMaxExecutionMS, minExecutionMS, issues)

	return Permissions{
		HostFunctions: hostFns,
		Network:       NetworkPermission{Allowed: netAllowed, AllowedHosts: netHosts},
		Filesystem:    FilesystemPermission{Allowed: fsAllowed, AllowedPaths: fsPaths},
		Env:           EnvPermission{Allowed: envAllowed, AllowedKeys: envKeys},
		MaxMemoryBytes: maxMem,
		MaxExecutionMS: maxExec,
	}
}

func parsePositiveInt(raw any, field string, fallback, minimum int64, issues *issueCollector) int64 {
	if raw == nil {
		return fallback
	}
	f, ok := raw.(float64)
	if !ok || f != float64(int64(f)) || int64(f) < minimum {
		issues.addUnique(fmt.Sprintf("%s must be an integer >= %d", field, minimum))
		return fallback
	}
	return int64(f)
}

func parseEntrypoint(raw any, issues *issueCollector) Entrypoint {
	rec, ok := raw.(map[string]any)
	if !ok {
		issues.addUnique("entrypoint must be an object")
		return Entrypoint{}
	}

	epType, _ := rec["type"].(string)
	switch EntrypointType(epType) {
	case EntrypointWASIComponent, EntrypointWASICommand, EntrypointCoreModule:
		// valid
	default:
		issues.addUnique("entrypoint.type must be one of wasi-component, wasi-command, core-module")
	}

	module, _ := rec["module"].(string)
	if !strings.HasSuffix(module, ".wasm") {
		issues.addUnique("entrypoint.module must be a path ending in .wasm")
	}

	var exports EntrypointExports
	if rawExports, ok := rec["exports"].(map[string]any); ok {
		if s, ok := rawExports["transform"].(string); ok && s != "" {
			exports.Transform = s
		} else if rawExports["transform"] != nil {
			issues.addUnique("entrypoint.exports.transform must be a non-empty string")
		}
		if s, ok := rawExports["hook"].(string); ok && s != "" {
			exports.Hook = s
		} else if rawExports["hook"] != nil {
			issues.addUnique("entrypoint.exports.hook must be a non-empty string")
		}
		if s, ok := rawExports["init"].(string); ok && s != "" {
			exports.Init = s
		} else if rawExports["init"] != nil {
			issues.addUnique("entrypoint.exports.init must be a non-empty string")
		}
	} else if rec["exports"] != nil {
		issues.addUnique("entrypoint.exports must be an object")
	}

	return Entrypoint{
		Type:    EntrypointType(epType),
		Module:  module,
		Exports: exports,
	}
}

func parseEngine(raw any, issues *issueCollector) Engine {
	rec, ok := raw.(map[string]any)
	if !ok {
		issues.addUnique("engine must be an object")
		return Engine{}
	}

	runtime, _ := rec["runtime"].(string)
	switch EngineRuntime(runtime) {
	case RuntimeWASIPreview1, RuntimeWASIPreview2, RuntimeCore:
		// valid
	default:
		issues.addUnique("engine.runtime must be one of wasi-preview1, wasi-preview2, core")
	}

	var minHostVersion string
	if rawMin, exists := rec["min_host_version"]; exists && rawMin != nil {
		if s, ok := rawMin.(string); ok {
			if !strictSemver.MatchString(s) {
				issues.addUnique("engine.min_host_version must be a MAJOR.MINOR.PATCH string")
			} else {
				minHostVersion = s
			}
		} else {
			issues.addUnique("engine.min_host_version must be a MAJOR.MINOR.PATCH string")
		}
	}

	return Engine{
		Runtime:        EngineRuntime(runtime),
		MinHostVersion: minHostVersion,
	}
}

// -----------------------------------------------------------------------
// Helpers
// -----------------------------------------------------------------------

func containsHostFn(fns []HostFunctionName, target HostFunctionName) bool {
	for _, fn := range fns {
		if fn == target {
			return true
		}
	}
	return false
}

// issueCollector accumulates unique validation issue strings.
type issueCollector struct {
	list []string
	set  map[string]struct{}
}

func (ic *issueCollector) add(msg string) {
	if ic.set == nil {
		ic.set = make(map[string]struct{})
	}
	ic.list = append(ic.list, msg)
	ic.set[msg] = struct{}{}
}

func (ic *issueCollector) addUnique(msg string) {
	if ic.set == nil {
		ic.set = make(map[string]struct{})
	}
	if _, exists := ic.set[msg]; !exists {
		ic.set[msg] = struct{}{}
		ic.list = append(ic.list, msg)
	}
}

File: packages/plugins/src/manifest.ts
import { ManifestValidationError } from './errors.ts';

/** JSON value type passed across the host/guest boundary. */
export type PluginJsonValue =
  | string
  | number
  | boolean
  | null
  | PluginJsonValue[]
  | { [key: string]: PluginJsonValue };

export const PLUGIN_SCHEMA_VERSION = 'ubag.plugin.v0' as const;

export type PluginCapability =
  | 'transform.prompt'
  | 'transform.response'
  | 'hook.job.pre'
  | 'hook.job.post'
  | 'hook.webhook.transform'
  | 'hook.validate'
  | 'adapter.extension'
  | 'command.custom';

export const PLUGIN_CAPABILITIES: readonly PluginCapability[] = [
  'transform.prompt',
  'transform.response',
  'hook.job.pre',
  'hook.job.post',
  'hook.webhook.transform',
  'hook.validate',
  'adapter.extension',
  'command.custom',
];

export type HostFunctionName = 'log' | 'clock' | 'random' | 'fetch' | 'read_file' | 'get_env';

export const HOST_FUNCTION_NAMES: readonly HostFunctionName[] = [
  'log',
  'clock',
  'random',
  'fetch',
  'read_file',
  'get_env',
];

export type EntrypointType = 'wasi-component' | 'wasi-command' | 'core-module';
export type EngineRuntime = 'wasi-preview1' | 'wasi-preview2' | 'core';

export interface PluginEntrypoint {
  readonly type: EntrypointType;
  readonly module: string;
  readonly exports: {
    readonly transform?: string;
    readonly hook?: string;
    readonly init?: string;
  };
}

export interface NetworkPermission {
  readonly allowed: boolean;
  readonly allowed_hosts: readonly string[];
}

export interface FilesystemPermission {
  readonly allowed: boolean;
  readonly allowed_paths: readonly string[];
}

export interface EnvPermission {
  readonly allowed: boolean;
  readonly allowed_keys: readonly string[];
}

export interface PluginPermissions {
  readonly host_functions: readonly HostFunctionName[];
  readonly network: NetworkPermission;
  readonly filesystem: FilesystemPermission;
  readonly env: EnvPermission;
  readonly max_memory_bytes: number;
  readonly max_execution_ms: number;
}

export interface PluginEngine {
  readonly runtime: EngineRuntime;
  readonly min_host_version: string | undefined;
}

export interface PluginManifest {
  readonly schema_version: typeof PLUGIN_SCHEMA_VERSION;
  readonly id: string;
  readonly display_name: string;
  readonly version: string;
  readonly description: string | undefined;
  readonly entrypoint: PluginEntrypoint;
  readonly capabilities: readonly PluginCapability[];
  readonly permissions: PluginPermissions;
  readonly engine: PluginEngine;
}

const ID_PATTERN = /^[a-z0-9][a-z0-9_-]{1,63}$/;
const SEMVER_PATTERN = /^[0-9]+\.[0-9]+\.[0-9]+([-+].+)?$/;
const STRICT_SEMVER_PATTERN = /^[0-9]+\.[0-9]+\.[0-9]+$/;
const DEFAULT_MAX_MEMORY = 16_777_216;
const DEFAULT_MAX_EXECUTION_MS = 1_000;

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function pushUnique(issues: string[], message: string): void {
  if (!issues.includes(message)) {
    issues.push(message);
  }
}

function parsePermissionToggle(
  raw: unknown,
  field: string,
  listKey: string,
  issues: string[],
): { allowed: boolean; list: readonly string[] } {
  if (raw === undefined) {
    return { allowed: false, list: [] };
  }
  if (!isRecord(raw)) {
    pushUnique(issues, `permissions.${field} must be an object`);
    return { allowed: false, list: [] };
  }
  const allowed = raw['allowed'] === true;
  if (typeof raw['allowed'] !== 'boolean') {
    pushUnique(issues, `permissions.${field}.allowed must be a boolean`);
  }
  const list: string[] = [];
  const rawList = raw[listKey];
  if (rawList !== undefined) {
    if (!Array.isArray(rawList) || rawList.some((entry) => typeof entry !== 'string')) {
      pushUnique(issues, `permissions.${field}.${listKey} must be a string array`);
    } else {
      list.push(...(rawList as string[]));
    }
  }
  if (!allowed && list.length > 0) {
    pushUnique(
      issues,
      `permissions.${field}.${listKey} is set but permissions.${field}.allowed is false`,
    );
  }
  return { allowed, list };
}

function parsePermissions(raw: unknown, issues: string[]): PluginPermissions {
  if (!isRecord(raw)) {
    pushUnique(issues, 'permissions must be an object');
    raw = {};
  }
  const record = raw as Record<string, unknown>;

  const hostFunctions: HostFunctionName[] = [];
  const rawHostFns = record['host_functions'];
  if (!Array.isArray(rawHostFns)) {
    pushUnique(issues, 'permissions.host_functions must be an array');
  } else {
    for (const entry of rawHostFns) {
      if (typeof entry !== 'string' || !HOST_FUNCTION_NAMES.includes(entry as HostFunctionName)) {
        pushUnique(issues, `permissions.host_functions contains unknown host function "${String(entry)}"`);
        continue;
      }
      if (!hostFunctions.includes(entry as HostFunctionName)) {
        hostFunctions.push(entry as HostFunctionName);
      }
    }
  }

  const network = parsePermissionToggle(record['network'], 'network', 'allowed_hosts', issues);
  const filesystem = parsePermissionToggle(record['filesystem'], 'filesystem', 'allowed_paths', issues);
  const env = parsePermissionToggle(record['env'], 'env', 'allowed_keys', issues);

  if (network.allowed && !hostFunctions.includes('fetch')) {
    pushUnique(issues, 'permissions.network.allowed requires host function "fetch"');
  }
  if (filesystem.allowed && !hostFunctions.includes('read_file')) {
    pushUnique(issues, 'permissions.filesystem.allowed requires host function "read_file"');
  }
  if (env.allowed && !hostFunctions.includes('get_env')) {
    pushUnique(issues, 'permissions.env.allowed requires host function "get_env"');
  }

  const maxMemory = parsePositiveInteger(
    record['max_memory_bytes'],
    'permissions.max_memory_bytes',
    DEFAULT_MAX_MEMORY,
    65_536,
    issues,
  );
  const maxExecution = parsePositiveInteger(
    record['max_execution_ms'],
    'permissions.max_execution_ms',
    DEFAULT_MAX_EXECUTION_MS,
    1,
    issues,
  );

  return {
    host_functions: hostFunctions,
    network: { allowed: network.allowed, allowed_hosts: network.list },
    filesystem: { allowed: filesystem.allowed, allowed_paths: filesystem.list },
    env: { allowed: env.allowed, allowed_keys: env.list },
    max_memory_bytes: maxMemory,
    max_execution_ms: maxExecution,
  };
}

function parsePositiveInteger(
  raw: unknown,
  field: string,
  fallback: number,
  minimum: number,
  issues: string[],
): number {
  if (raw === undefined) {
    return fallback;
  }
  if (typeof raw !== 'number' || !Number.isInteger(raw) || raw < minimum) {
    pushUnique(issues, `${field} must be an integer >= ${minimum}`);
    return fallback;
  }
  return raw;
}

function parseEntrypoint(raw: unknown, issues: string[]): PluginEntrypoint {
  if (!isRecord(raw)) {
    pushUnique(issues, 'entrypoint must be an object');
    return { type: 'wasi-component', module: '', exports: {} };
  }
  const type = raw['type'];
  if (type !== 'wasi-component' && type !== 'wasi-command' && type !== 'core-module') {
    pushUnique(issues, 'entrypoint.type must be one of wasi-component, wasi-command, core-module');
  }
  const moduleValue = raw['module'];
  if (typeof moduleValue !== 'string' || !moduleValue.endsWith('.wasm')) {
    pushUnique(issues, 'entrypoint.module must be a path ending in .wasm');
  }
  const exportsRaw = raw['exports'];
  const exportsOut: { transform?: string; hook?: string; init?: string } = {};
  if (!isRecord(exportsRaw)) {
    pushUnique(issues, 'entrypoint.exports must be an object');
  } else {
    for (const key of ['transform', 'hook', 'init'] as const) {
      const value = exportsRaw[key];
      if (value === undefined) {
        continue;
      }
      if (typeof value !== 'string' || value.length === 0) {
        pushUnique(issues, `entrypoint.exports.${key} must be a non-empty string`);
        continue;
      }
      exportsOut[key] = value;
    }
  }
  return {
    type: (type as EntrypointType) ?? 'wasi-component',
    module: typeof moduleValue === 'string' ? moduleValue : '',
    exports: exportsOut,
  };
}

function parseEngine(raw: unknown, issues: string[]): PluginEngine {
  if (!isRecord(raw)) {
    pushUnique(issues, 'engine must be an object');
    return { runtime: 'wasi-preview2', min_host_version: undefined };
  }
  const runtime = raw['runtime'];
  if (runtime !== 'wasi-preview1' && runtime !== 'wasi-preview2' && runtime !== 'core') {
    pushUnique(issues, 'engine.runtime must be one of wasi-preview1, wasi-preview2, core');
  }
  let minHostVersion: string | undefined;
  const rawMin = raw['min_host_version'];
  if (rawMin !== undefined) {
    if (typeof rawMin !== 'string' || !STRICT_SEMVER_PATTERN.test(rawMin)) {
      pushUnique(issues, 'engine.min_host_version must be a MAJOR.MINOR.PATCH string');
    } else {
      minHostVersion = rawMin;
    }
  }
  return {
    runtime: (runtime as EngineRuntime) ?? 'wasi-preview2',
    min_host_version: minHostVersion,
  };
}

/**
 * Validate and normalize an untrusted manifest object.
 *
 * @throws ManifestValidationError when the manifest violates the schema.
 */
export function parsePluginManifest(input: unknown): PluginManifest {
  const issues: string[] = [];

  if (!isRecord(input)) {
    throw new ManifestValidationError(['manifest must be a JSON object']);
  }

  if (input['schema_version'] !== PLUGIN_SCHEMA_VERSION) {
    pushUnique(issues, `schema_version must be "${PLUGIN_SCHEMA_VERSION}"`);
  }

  const id = input['id'];
  if (typeof id !== 'string' || !ID_PATTERN.test(id)) {
    pushUnique(issues, 'id must match ^[a-z0-9][a-z0-9_-]{1,63}$');
  }

  const displayName = input['display_name'];
  if (typeof displayName !== 'string' || displayName.length === 0) {
    pushUnique(issues, 'display_name must be a non-empty string');
  }

  const version = input['version'];
  if (typeof version !== 'string' || !SEMVER_PATTERN.test(version)) {
    pushUnique(issues, 'version must be a semantic version string');
  }

  let description: string | undefined;
  const rawDescription = input['description'];
  if (rawDescription !== undefined) {
    if (typeof rawDescription !== 'string') {
      pushUnique(issues, 'description must be a string');
    } else {
      description = rawDescription;
    }
  }

  const capabilities: PluginCapability[] = [];
  const rawCapabilities = input['capabilities'];
  if (!Array.isArray(rawCapabilities) || rawCapabilities.length === 0) {
    pushUnique(issues, 'capabilities must be a non-empty array');
  } else {
    for (const entry of rawCapabilities) {
      if (typeof entry !== 'string' || !PLUGIN_CAPABILITIES.includes(entry as PluginCapability)) {
        pushUnique(issues, `capabilities contains unknown capability "${String(entry)}"`);
        continue;
      }
      if (!capabilities.includes(entry as PluginCapability)) {
        capabilities.push(entry as PluginCapability);
      }
    }
  }

  const entrypoint = parseEntrypoint(input['entrypoint'], issues);
  const permissions = parsePermissions(input['permissions'], issues);
  const engine = parseEngine(input['engine'], issues);

  if (issues.length > 0) {
    throw new ManifestValidationError(issues, typeof id === 'string' ? id : undefined);
  }

  return {
    schema_version: PLUGIN_SCHEMA_VERSION,
    id: id as string,
    display_name: displayName as string,
    version: version as string,
    description,
    entrypoint,
    capabilities,
    permissions,
    engine,
  };
}

/** Parse manifest JSON text. */
export function parsePluginManifestJson(text: string): PluginManifest {
  let parsed: unknown;
  try {
    parsed = JSON.parse(text);
  } catch (error) {
    throw new ManifestValidationError([
      `manifest is not valid JSON: ${(error as Error).message}`,
    ]);
  }
  return parsePluginManifest(parsed);
}
